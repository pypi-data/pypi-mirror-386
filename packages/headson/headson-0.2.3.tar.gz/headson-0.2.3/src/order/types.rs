use serde_json;

#[derive(Clone, Debug)]
pub struct PriorityConfig {
    pub max_string_graphemes: usize,
    pub array_max_items: usize,
}

impl PriorityConfig {
    pub fn new(max_string_graphemes: usize, array_max_items: usize) -> Self {
        Self {
            max_string_graphemes,
            array_max_items,
        }
    }
}

#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
pub struct NodeId(pub usize);

#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
pub enum NodeKind {
    Null,
    Bool,
    Number,
    String,
    Array,
    Object,
}

#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
pub enum ObjectType {
    Object,
    Fileset,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct RankedNode {
    pub node_id: NodeId,
    pub kind: NodeKind,
    pub key_in_object: Option<String>,
    pub number_value: Option<serde_json::Number>,
    pub bool_value: Option<bool>,
    pub string_value: Option<String>,
}

#[derive(Clone, Debug, Default)]
pub struct NodeMetrics {
    pub array_len: Option<usize>,
    pub object_len: Option<usize>,
    pub string_len: Option<usize>,
    pub string_truncated: bool,
}

#[derive(Clone, Debug)]
pub struct PriorityOrder {
    pub metrics: Vec<NodeMetrics>,
    pub nodes: Vec<RankedNode>,
    // All ids in this structure are PQ ids (0..total_nodes).
    // They correspond to `NodeId.0` in `RankedNode` for convenience when indexing.
    pub parent: Vec<Option<NodeId>>, // parent[id] = parent id (PQ id)
    pub children: Vec<Vec<NodeId>>,  // children[id] = children ids (PQ ids)
    pub order: Vec<NodeId>, // ids sorted by ascending priority (PQ ids)
    pub total_nodes: usize,
    pub object_type: Vec<ObjectType>,
}

pub const ROOT_PQ_ID: usize = 0;
