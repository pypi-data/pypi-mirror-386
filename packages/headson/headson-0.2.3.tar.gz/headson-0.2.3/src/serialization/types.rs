#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum OutputTemplate {
    Json,
    Pseudo,
    Js,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct RenderConfig {
    pub template: OutputTemplate,
    pub indent_unit: String,
    pub space: String,
    // Newline sequence to use in final output (e.g., "\n" or "").
    // Templates read this directly; no post-processing replacement.
    pub newline: String,
}
