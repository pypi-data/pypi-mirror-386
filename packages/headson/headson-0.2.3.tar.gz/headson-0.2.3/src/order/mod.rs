pub mod build;
pub mod scoring;
pub mod types;

pub use build::build_order;
pub use types::{
    NodeId, NodeKind, PriorityConfig, PriorityOrder, ROOT_PQ_ID, RankedNode,
};
