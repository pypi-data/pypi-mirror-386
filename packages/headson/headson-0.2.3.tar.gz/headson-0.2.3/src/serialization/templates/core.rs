use super::super::indent;
use super::{ArrayCtx, ObjectCtx};

// Shared rendering core for all templates.
// - Style controls only empty/omitted decorations.
// - Indentation and newlines come from ctx (depth, indent_unit, newline).
// - When ctx.inline_open is true, no leading indent is emitted before the opener.
pub trait Style {
    fn array_empty(open_indent: &str, ctx: &ArrayCtx<'_>) -> String;
    fn array_push_omitted(_out: &mut String, _ctx: &ArrayCtx<'_>) {}

    fn object_empty(open_indent: &str, ctx: &ObjectCtx<'_>) -> String;
    fn object_push_omitted(_out: &mut String, _ctx: &ObjectCtx<'_>) {}
}

fn push_array_items(out: &mut String, ctx: &ArrayCtx<'_>) {
    for (i, (_, item)) in ctx.children.iter().enumerate() {
        out.push_str(item);
        if i + 1 < ctx.children_len {
            out.push(',');
        }
        out.push_str(ctx.newline);
    }
}

fn push_object_items(out: &mut String, ctx: &ObjectCtx<'_>) {
    for (i, (_, (k, v))) in ctx.children.iter().enumerate() {
        out.push_str(&indent(ctx.depth + 1, ctx.indent_unit));
        out.push_str(k);
        out.push(':');
        out.push_str(ctx.space);
        out.push_str(v);
        if i + 1 < ctx.children_len {
            out.push(',');
        }
        out.push_str(ctx.newline);
    }
}

// Render an array using the shared control flow and style-specific decorations.
pub fn render_array_with<S: Style>(ctx: &ArrayCtx<'_>) -> String {
    let base = indent(ctx.depth, ctx.indent_unit);
    let open_indent = if ctx.inline_open { "" } else { &base };
    if ctx.children_len == 0 {
        return S::array_empty(open_indent, ctx);
    }
    let mut out = String::new();
    out.push_str(open_indent);
    out.push('[');
    out.push_str(ctx.newline);
    push_array_items(&mut out, ctx);
    S::array_push_omitted(&mut out, ctx);
    out.push_str(&base);
    out.push(']');
    out
}

// Render an object using the shared control flow and style-specific decorations.
pub fn render_object_with<S: Style>(ctx: &ObjectCtx<'_>) -> String {
    let base = indent(ctx.depth, ctx.indent_unit);
    let open_indent = if ctx.inline_open { "" } else { &base };
    if ctx.children_len == 0 {
        return S::object_empty(open_indent, ctx);
    }
    let mut out = String::new();
    out.push_str(open_indent);
    out.push('{');
    out.push_str(ctx.newline);
    push_object_items(&mut out, ctx);
    S::object_push_omitted(&mut out, ctx);
    out.push_str(&base);
    out.push('}');
    out
}
