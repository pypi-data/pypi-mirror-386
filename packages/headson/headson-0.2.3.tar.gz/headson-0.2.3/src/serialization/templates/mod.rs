use crate::OutputTemplate;

pub mod core;
pub mod js;
pub mod json;
pub mod pseudo;

pub struct ArrayCtx<'a> {
    pub children: Vec<(usize, String)>,
    pub children_len: usize,
    pub omitted: usize,
    pub depth: usize,
    pub indent_unit: &'a str,
    pub inline_open: bool,
    pub newline: &'a str,
}

pub struct ObjectCtx<'a> {
    pub children: Vec<(usize, (String, String))>,
    pub children_len: usize,
    pub omitted: usize,
    pub depth: usize,
    pub indent_unit: &'a str,
    pub inline_open: bool,
    pub space: &'a str,
    pub newline: &'a str,
    pub fileset_root: bool,
}

pub fn render_array(template: OutputTemplate, ctx: &ArrayCtx<'_>) -> String {
    match template {
        OutputTemplate::Json => json::render_array(ctx),
        OutputTemplate::Pseudo => pseudo::render_array(ctx),
        OutputTemplate::Js => js::render_array(ctx),
    }
}

pub fn render_object(template: OutputTemplate, ctx: &ObjectCtx<'_>) -> String {
    match template {
        OutputTemplate::Json => json::render_object(ctx),
        OutputTemplate::Pseudo => pseudo::render_object(ctx),
        OutputTemplate::Js => js::render_object(ctx),
    }
}
