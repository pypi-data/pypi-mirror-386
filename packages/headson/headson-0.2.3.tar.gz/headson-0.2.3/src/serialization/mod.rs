use crate::order::types::ObjectType;
use crate::order::{NodeKind, PriorityOrder, ROOT_PQ_ID};
pub mod templates;
pub mod types;
use self::templates::{ArrayCtx, ObjectCtx, render_array, render_object};

fn indent(depth: usize, unit: &str) -> String {
    unit.repeat(depth)
}

type ArrayChildPair = (usize, String);
type ObjectChildPair = (usize, (String, String));

pub(crate) struct RenderScope<'a> {
    pq: &'a PriorityOrder,
    marks: &'a [u32],
    mark_gen: u32,
    config: &'a crate::RenderConfig,
}

impl<'a> RenderScope<'a> {
    fn append_js_fileset_section(
        &mut self,
        out: &mut String,
        depth: usize,
        child_pq_id: usize,
        nl: &str,
    ) {
        let raw_key = self.pq.nodes[child_pq_id]
            .key_in_object
            .as_deref()
            .unwrap_or("");
        out.push_str(&indent(depth, &self.config.indent_unit));
        out.push_str("// ");
        out.push_str(raw_key);
        out.push_str(nl);
        let rendered = self.serialize_node(child_pq_id, depth, false);
        out.push_str(&rendered);
        out.push(';');
        out.push_str(nl);
    }

    fn append_js_fileset_summary(
        &self,
        out: &mut String,
        depth: usize,
        kept: usize,
        total: usize,
        nl: &str,
    ) {
        if total > kept && !nl.is_empty() {
            let blanks = if out.ends_with(nl) { 1 } else { 2 };
            for _ in 0..blanks {
                out.push_str(nl);
            }
            out.push_str(&indent(depth, &self.config.indent_unit));
            out.push_str(&format!("/* {} more files */", total - kept));
            out.push_str(nl);
        }
    }

    fn append_pseudo_fileset_section(
        &mut self,
        out: &mut String,
        depth: usize,
        child_pq_id: usize,
        nl: &str,
    ) {
        let raw_key = self.pq.nodes[child_pq_id]
            .key_in_object
            .as_deref()
            .unwrap_or("");
        out.push_str(&indent(depth, &self.config.indent_unit));
        out.push_str("==> ");
        out.push_str(raw_key);
        out.push_str(" <==");
        out.push_str(nl);
        let rendered = self.serialize_node(child_pq_id, depth, false);
        out.push_str(&rendered);
    }

    fn append_pseudo_fileset_summary(
        &self,
        out: &mut String,
        depth: usize,
        kept: usize,
        total: usize,
        nl: &str,
    ) {
        if total > kept && !nl.is_empty() {
            while out.ends_with(nl) {
                let new_len = out.len().saturating_sub(nl.len());
                out.truncate(new_len);
            }
            out.push_str(nl);
            out.push_str(nl);
            out.push_str(&indent(depth, &self.config.indent_unit));
            out.push_str(&format!("==> {} more files <==", total - kept));
        }
    }
    fn count_kept_children(&self, id: usize) -> usize {
        if let Some(kids) = self.pq.children.get(id) {
            let mut kept = 0usize;
            for &cid in kids {
                if self.marks[cid.0] == self.mark_gen {
                    kept += 1;
                }
            }
            kept
        } else {
            0
        }
    }

    fn omitted_for_string(&self, id: usize, kept: usize) -> Option<usize> {
        let m = &self.pq.metrics[id];
        if let Some(orig) = m.string_len {
            if orig > kept {
                return Some(orig - kept);
            }
            if m.string_truncated {
                return Some(1);
            }
            None
        } else if m.string_truncated {
            Some(1)
        } else {
            None
        }
    }

    fn omitted_for(
        &self,
        id: usize,
        kind: NodeKind,
        kept: usize,
    ) -> Option<usize> {
        match kind {
            NodeKind::Array => {
                self.pq.metrics[id].array_len.and_then(|orig| {
                    if orig > kept { Some(orig - kept) } else { None }
                })
            }
            NodeKind::String => self.omitted_for_string(id, kept),
            NodeKind::Object => {
                self.pq.metrics[id].object_len.and_then(|orig| {
                    if orig > kept { Some(orig - kept) } else { None }
                })
            }
            _ => None,
        }
    }

    fn serialize_array(
        &mut self,
        id: usize,
        depth: usize,
        inline: bool,
    ) -> String {
        let config = self.config;
        let (children_pairs, kept) = self.gather_array_children(id, depth);
        let node = &self.pq.nodes[id];
        let omitted = self.omitted_for(id, node.kind, kept).unwrap_or(0);
        if kept == 0 && omitted == 0 {
            return "[]".to_string();
        }
        let ctx = ArrayCtx {
            children: children_pairs,
            children_len: kept,
            omitted,
            depth,
            indent_unit: &config.indent_unit,
            inline_open: inline,
            newline: &config.newline,
        };
        render_array(config.template, &ctx)
    }

    fn serialize_object(
        &mut self,
        id: usize,
        depth: usize,
        inline: bool,
    ) -> String {
        let config = self.config;
        // Special-case: fileset root in Pseudo/JS templates → head-style sections
        if id == ROOT_PQ_ID
            && self.pq.object_type.get(id) == Some(&ObjectType::Fileset)
            && !config.newline.is_empty()
        {
            match config.template {
                crate::OutputTemplate::Pseudo => {
                    return self.serialize_fileset_root_pseudo(depth);
                }
                crate::OutputTemplate::Js => {
                    return self.serialize_fileset_root_js(depth);
                }
                _ => {}
            }
        }
        let (children_pairs, kept) = self.gather_object_children(id, depth);
        let node = &self.pq.nodes[id];
        let omitted = self.omitted_for(id, node.kind, kept).unwrap_or(0);
        if kept == 0 && omitted == 0 {
            return "{}".to_string();
        }
        let ctx = ObjectCtx {
            children: children_pairs,
            children_len: kept,
            omitted,
            depth,
            indent_unit: &config.indent_unit,
            inline_open: inline,
            space: &config.space,
            newline: &config.newline,
            fileset_root: id == ROOT_PQ_ID
                && self.pq.object_type.get(id) == Some(&ObjectType::Fileset),
        };
        render_object(config.template, &ctx)
    }

    fn serialize_string(&mut self, id: usize) -> String {
        let kept = self.count_kept_children(id);
        let node = &self.pq.nodes[id];
        let omitted = self.omitted_for(id, node.kind, kept).unwrap_or(0);
        let full: &str = node.string_value.as_deref().unwrap_or("");
        if omitted == 0 {
            return crate::utils::json::json_string(full);
        }
        let prefix = crate::utils::text::take_n_graphemes(full, kept);
        let truncated = format!("{prefix}…");
        crate::utils::json::json_string(&truncated)
    }

    fn serialize_number(&self, id: usize) -> String {
        let it = &self.pq.nodes[id];
        if let Some(n) = it.number_value.as_ref() {
            if let Some(i) = n.as_i64() {
                return i.to_string();
            }
            if let Some(u) = n.as_u64() {
                return u.to_string();
            }
            if n.as_f64().is_some() {
                return n.to_string();
            }
        }
        "0".to_string()
    }

    fn serialize_bool(&self, id: usize) -> String {
        let it = &self.pq.nodes[id];
        match it.bool_value {
            Some(true) => "true".to_string(),
            Some(false) | None => "false".to_string(),
        }
    }

    fn serialize_node(
        &mut self,
        id: usize,
        depth: usize,
        inline: bool,
    ) -> String {
        let it = &self.pq.nodes[id];
        match it.kind {
            NodeKind::Array => self.serialize_array(id, depth, inline),
            NodeKind::Object => self.serialize_object(id, depth, inline),
            NodeKind::String => self.serialize_string(id),
            NodeKind::Number => self.serialize_number(id),
            NodeKind::Bool => self.serialize_bool(id),
            NodeKind::Null => "null".to_string(),
        }
    }

    fn gather_array_children(
        &mut self,
        id: usize,
        depth: usize,
    ) -> (Vec<ArrayChildPair>, usize) {
        let config = self.config;
        let mut children_pairs: Vec<ArrayChildPair> = Vec::new();
        let mut kept = 0usize;
        if let Some(children_ids) = self.pq.children.get(id) {
            for (i, &child_id) in children_ids.iter().enumerate() {
                if self.marks[child_id.0] != self.mark_gen {
                    continue;
                }
                kept += 1;
                let rendered =
                    self.serialize_node(child_id.0, depth + 1, false);
                if !config.newline.is_empty()
                    && rendered.contains(&config.newline)
                {
                    children_pairs.push((i, rendered));
                } else {
                    let child_indent = indent(depth + 1, &config.indent_unit);
                    children_pairs
                        .push((i, format!("{child_indent}{rendered}")));
                }
            }
        }
        (children_pairs, kept)
    }

    fn gather_object_children(
        &mut self,
        id: usize,
        depth: usize,
    ) -> (Vec<ObjectChildPair>, usize) {
        let mut children_pairs: Vec<ObjectChildPair> = Vec::new();
        let mut kept = 0usize;
        if let Some(children_ids) = self.pq.children.get(id) {
            for (i, &child_id) in children_ids.iter().enumerate() {
                if self.marks[child_id.0] != self.mark_gen {
                    continue;
                }
                kept += 1;
                let child = &self.pq.nodes[child_id.0];
                let raw_key = child.key_in_object.as_deref().unwrap_or("");
                let key = crate::utils::json::json_string(raw_key);
                let val = self.serialize_node(child_id.0, depth + 1, true);
                children_pairs.push((i, (key, val)));
            }
        }
        (children_pairs, kept)
    }

    // Render multi-input fileset as head-style sections for Pseudo template.
    fn serialize_fileset_root_pseudo(&mut self, depth: usize) -> String {
        let nl = &self.config.newline;
        let mut out = String::new();
        if let Some(children_ids) = self.pq.children.get(ROOT_PQ_ID) {
            let mut kept = 0usize;
            for &child_id in children_ids.iter() {
                if self.marks[child_id.0] != self.mark_gen {
                    continue;
                }
                if kept > 0 {
                    // ensure an empty line between files regardless of whether
                    // previous section ended with a newline
                    out.push_str(nl);
                    out.push_str(nl);
                }
                kept += 1;
                self.append_pseudo_fileset_section(
                    &mut out, depth, child_id.0, nl,
                );
            }
            let total = self
                .pq
                .metrics
                .get(ROOT_PQ_ID)
                .and_then(|m| m.object_len)
                .unwrap_or(children_ids.len());
            self.append_pseudo_fileset_summary(
                &mut out, depth, kept, total, nl,
            );
        }
        out
    }

    // Render multi-input fileset as head-style sections for JS template using
    // line comments for file names. Ensure valid JS by terminating each section
    // with a semicolon.
    fn serialize_fileset_root_js(&mut self, depth: usize) -> String {
        let nl = &self.config.newline;
        let mut out = String::new();
        let Some(children_ids) = self.pq.children.get(ROOT_PQ_ID) else {
            return out;
        };
        let kept =
            self.render_js_fileset_sections(&mut out, depth, children_ids, nl);
        let total = self
            .pq
            .metrics
            .get(ROOT_PQ_ID)
            .and_then(|m| m.object_len)
            .unwrap_or(children_ids.len());
        self.append_js_fileset_summary(&mut out, depth, kept, total, nl);
        out
    }

    fn render_js_fileset_sections(
        &mut self,
        out: &mut String,
        depth: usize,
        children_ids: &[crate::order::NodeId],
        nl: &str,
    ) -> usize {
        let mut kept = 0usize;
        for &child_id in children_ids.iter() {
            if self.marks[child_id.0] != self.mark_gen {
                continue;
            }
            if kept > 0 {
                out.push_str(nl);
            }
            kept += 1;
            self.append_js_fileset_section(out, depth, child_id.0, nl);
        }
        kept
    }
}

/// Render a budget-limited preview directly from the arena using inclusion marks.
pub fn render_arena_with_marks(
    order_build: &PriorityOrder,
    budget: usize,
    marks: &mut Vec<u32>,
    mark_gen: u32,
    config: &crate::RenderConfig,
) -> String {
    if marks.len() < order_build.total_nodes {
        marks.resize(order_build.total_nodes, 0);
    }
    // Phase 1: Mark the first `k` nodes (order[..k]) and all their ancestors
    let k = budget.min(order_build.total_nodes);
    crate::utils::graph::mark_top_k_and_ancestors(
        order_build,
        k,
        marks,
        mark_gen,
    );

    // Root PQ id is a fixed invariant (0).
    let root_id = ROOT_PQ_ID;
    let mut scope = RenderScope {
        pq: order_build,
        marks,
        mark_gen,
        config,
    };
    scope.serialize_node(root_id, 0, false)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::order::build_order;
    use insta::assert_snapshot;

    #[test]
    fn arena_render_empty_array() {
        let arena = crate::json_ingest::build_json_tree_arena(
            "[]",
            &crate::PriorityConfig::new(usize::MAX, usize::MAX),
        )
        .unwrap();
        let build = build_order(
            &arena,
            &crate::PriorityConfig::new(usize::MAX, usize::MAX),
        )
        .unwrap();
        let mut marks = vec![0u32; build.total_nodes];
        let out = render_arena_with_marks(
            &build,
            10,
            &mut marks,
            1,
            &crate::RenderConfig {
                template: crate::OutputTemplate::Json,
                indent_unit: "  ".to_string(),
                space: " ".to_string(),
                newline: "\n".to_string(),
            },
        );
        assert_snapshot!("arena_render_empty", out);
    }

    #[test]
    fn arena_render_single_string_array() {
        let arena = crate::json_ingest::build_json_tree_arena(
            "[\"ab\"]",
            &crate::PriorityConfig::new(usize::MAX, usize::MAX),
        )
        .unwrap();
        let build = build_order(
            &arena,
            &crate::PriorityConfig::new(usize::MAX, usize::MAX),
        )
        .unwrap();
        let mut marks = vec![0u32; build.total_nodes];
        let out = render_arena_with_marks(
            &build,
            10,
            &mut marks,
            1,
            &crate::RenderConfig {
                template: crate::OutputTemplate::Json,
                indent_unit: "  ".to_string(),
                space: " ".to_string(),
                newline: "\n".to_string(),
            },
        );
        assert_snapshot!("arena_render_single", out);
    }
}
