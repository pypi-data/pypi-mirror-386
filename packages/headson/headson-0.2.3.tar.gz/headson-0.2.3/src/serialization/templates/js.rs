use super::super::indent;
use super::core::{Style, render_array_with, render_object_with};
use super::{ArrayCtx, ObjectCtx};

struct Js;

impl Style for Js {
    fn array_empty(open_indent: &str, ctx: &ArrayCtx<'_>) -> String {
        if ctx.omitted > 0 {
            return format!(
                "{open_indent}[ /* {} more items */ ]",
                ctx.omitted
            );
        }
        format!("{open_indent}[ /* empty */ ]")
    }

    fn array_push_omitted(out: &mut String, ctx: &ArrayCtx<'_>) {
        if ctx.omitted > 0 {
            out.push_str(&indent(ctx.depth + 1, ctx.indent_unit));
            out.push_str(&format!(
                "/* {} more items */{}",
                ctx.omitted, ctx.newline
            ));
        }
    }

    fn object_empty(open_indent: &str, ctx: &ObjectCtx<'_>) -> String {
        if ctx.omitted > 0 {
            let label = if ctx.fileset_root {
                "files"
            } else {
                "properties"
            };
            return format!(
                "{open_indent}{{{space}/* {n} more {label} */{space}}}",
                n = ctx.omitted,
                space = ctx.space
            );
        }
        format!(
            "{open_indent}{{{space}/* empty */{space}}}",
            space = ctx.space
        )
    }

    fn object_push_omitted(out: &mut String, ctx: &ObjectCtx<'_>) {
        if ctx.omitted > 0 {
            out.push_str(&indent(ctx.depth + 1, ctx.indent_unit));
            let label = if ctx.fileset_root {
                "files"
            } else {
                "properties"
            };
            out.push_str(&format!(
                "/* {} more {label} */{}",
                ctx.omitted, ctx.newline
            ));
        }
    }
}

pub fn render_array(ctx: &ArrayCtx<'_>) -> String {
    render_array_with::<Js>(ctx)
}

pub fn render_object(ctx: &ObjectCtx<'_>) -> String {
    render_object_with::<Js>(ctx)
}
