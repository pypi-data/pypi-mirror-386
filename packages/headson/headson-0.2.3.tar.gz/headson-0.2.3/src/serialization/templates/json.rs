use super::core::{Style, render_array_with, render_object_with};
use super::{ArrayCtx, ObjectCtx};

struct Json;

impl Style for Json {
    fn array_empty(open_indent: &str, _ctx: &ArrayCtx<'_>) -> String {
        format!("{open_indent}[]")
    }

    fn object_empty(open_indent: &str, _ctx: &ObjectCtx<'_>) -> String {
        format!("{open_indent}{{}}")
    }
}

pub fn render_array(ctx: &ArrayCtx<'_>) -> String {
    render_array_with::<Json>(ctx)
}

pub fn render_object(ctx: &ObjectCtx<'_>) -> String {
    render_object_with::<Json>(ctx)
}
