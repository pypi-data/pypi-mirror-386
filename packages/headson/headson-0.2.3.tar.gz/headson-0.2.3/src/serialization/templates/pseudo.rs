use super::super::indent;
use super::core::{Style, render_array_with, render_object_with};
use super::{ArrayCtx, ObjectCtx};

struct Pseudo;

impl Style for Pseudo {
    fn array_empty(open_indent: &str, ctx: &ArrayCtx<'_>) -> String {
        if ctx.omitted > 0 {
            return format!("{open_indent}[ … ]");
        }
        format!("{open_indent}[]")
    }

    fn array_push_omitted(out: &mut String, ctx: &ArrayCtx<'_>) {
        if ctx.omitted > 0 {
            out.push_str(&indent(ctx.depth + 1, ctx.indent_unit));
            out.push('…');
            out.push_str(ctx.newline);
        }
    }

    fn object_empty(open_indent: &str, ctx: &ObjectCtx<'_>) -> String {
        if ctx.omitted > 0 {
            return format!(
                "{open_indent}{{{space}…{space}}}",
                space = ctx.space
            );
        }
        format!("{open_indent}{{}}")
    }

    fn object_push_omitted(out: &mut String, ctx: &ObjectCtx<'_>) {
        if ctx.omitted > 0 {
            out.push_str(&indent(ctx.depth + 1, ctx.indent_unit));
            out.push('…');
            out.push_str(ctx.newline);
        }
    }
}

pub fn render_array(ctx: &ArrayCtx<'_>) -> String {
    render_array_with::<Pseudo>(ctx)
}

pub fn render_object(ctx: &ObjectCtx<'_>) -> String {
    render_object_with::<Pseudo>(ctx)
}
