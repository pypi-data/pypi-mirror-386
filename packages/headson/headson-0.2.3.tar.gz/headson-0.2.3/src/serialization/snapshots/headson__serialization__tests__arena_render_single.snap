---
source: src/serialization/mod.rs
expression: out
---
[
  "ab"
]
